src/language/dafnyInstallation.ts

```diff
@@ -18,6 +18,47 @@ export class DafnyInstaller {
     this.customPathInstaller = new CustomPathInstaller(this.context, this.statusOutput);
   }
 
+
+  public static readonly CurrentVersionTag = 'current-version';
+
+  // If there is a major version bump and the version is auto-selected
+  // warn the user before hand
+  public static async dafny4upgradeCheck(
+    context: ExtensionContext,
+    versionDescription: string,
+    newToolVersion: string
+  ): Promise<string> {
+    const previousToolVersion: string | undefined = context.globalState.get(DafnyInstaller.CurrentVersionTag);
+    // In case users have not reopened VSCode since this code was added until Dafny 4 is released
+    const compareToolVersion: string = previousToolVersion ?? '3.10.0';
+    if(!newToolVersion.startsWith(compareToolVersion[0])) {
+      const seeMigrationInstructions = 'Show migration instructions';
+      const accept = `Accept ${newToolVersion}`;
+      const decline = `Keep ${compareToolVersion} for now`;
+      let answer: string | undefined = seeMigrationInstructions;
+
+      while(answer === seeMigrationInstructions) {
+        // Major version change
+        answer = await window.showInformationMessage(
+          `Dafny ${newToolVersion} is out! You are using the ${versionDescription.toLowerCase()}, `
+          + 'but because this is a major change, we want you to confirm the switch.',
+          seeMigrationInstructions,
+          accept,
+          decline
+        );
+        if(answer === undefined || answer === decline) {
+          newToolVersion = (compareToolVersion !== undefined ? compareToolVersion : newToolVersion) as string;
+        } else if(answer === seeMigrationInstructions) {
+          env.openExternal(Uri.parse('https://dafny.org/TODO'));
```

Comment:
Why not refer to the release notes?

Reply:
Our release notes won't be detailed enough for that purpose. We need a better migration document. Perhaps @davidcok  knows where would be a good place to start such a document?